lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs

```diff
@@ -0,0 +1,727 @@
+        let req = ServerDeviceAnnounceResponse::decode(payload)?;
+        debug!("got {:?}", req);
+
+        if req.device_id != SCARD_DEVICE_ID {
+            Err(invalid_data_error(&format!(
+                "got ServerDeviceAnnounceResponse for unknown device_id {}",
+                &req.device_id
+            )))
+        } else if req.result_code != NTSTATUS_OK {
+            Err(invalid_data_error(&format!(
+                "got unsuccessful ServerDeviceAnnounceResponse result code NTSTATUS({})",
+                &req.result_code
+            )))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn handle_device_io_request(&self, payload: &mut Payload) -> RdpResult<Option<Vec<u8>>> {
+        let req = DeviceIoRequest::decode(payload)?;
+        debug!("got {:?}", req);
+
+        if let MajorFunction::IRP_MJ_DEVICE_CONTROL = req.major_function {
+            let ioctl = DeviceControlRequest::decode(req, payload)?;
+            debug!("got {:?}", ioctl);
+
+            let (code, res) = self.scard.ioctl(ioctl.io_control_code, payload)?;
+            if code == SPECIAL_NO_RESPONSE {
+                return Ok(None);
+            }
+            let resp = encode_message(
+                PacketId::PAKID_CORE_DEVICE_IOCOMPLETION,
+                DeviceControlResponse::new(&ioctl, code, res).encode()?,
+            )?;
+            debug!("sending device IO response");
+            Ok(Some(resp))
+        } else {
+            Err(invalid_data_error(&format!(
+                "got unsupported major_function in DeviceIoRequest: {:?}",
+                &req.major_function
+            )))
+        }
+    }
+}
+
+impl Default for Client {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+fn encode_message(packet_id: PacketId, payload: Vec<u8>) -> RdpResult<Vec<u8>> {
+    let mut inner = Header::new(Component::RDPDR_CTYP_CORE, packet_id).encode()?;
+    inner.extend_from_slice(&payload);
+    let mut outer = ChannelPDUHeader::new(inner.length() as u32).encode()?;
+    outer.extend_from_slice(&inner);
+    Ok(outer)
+}
+
+bitflags! {
+    struct ChannelPDUFlags: u32 {
+        const CHANNEL_FLAG_FIRST = 0x00000001;
+        const CHANNEL_FLAG_LAST = 0x00000002;
+        const CHANNEL_FLAG_ONLY = Self::CHANNEL_FLAG_FIRST.bits | Self::CHANNEL_FLAG_LAST.bits;
+    }
+}
+
+#[derive(Debug)]
+struct ChannelPDUHeader {
+    length: u32,
+    flags: ChannelPDUFlags,
+}
+
+impl ChannelPDUHeader {
+    fn new(length: u32) -> Self {
+        Self {
+            length,
+            flags: ChannelPDUFlags::CHANNEL_FLAG_ONLY,
+        }
+    }
+    fn decode(payload: &mut Payload) -> RdpResult<Self> {
+        Ok(Self {
+            length: payload.read_u32::<LittleEndian>()?,
+            flags: ChannelPDUFlags::from_bits(payload.read_u32::<LittleEndian>()?)
+                .ok_or_else(|| invalid_data_error("invalid flags in ChannelPDUHeader"))?,
+        })
+    }
+    fn encode(&self) -> RdpResult<Vec<u8>> {
```

Comment:
Note: Its generally more "rusty" to avoid intermediate allocations and pass in a mutable buffer/collector as an argument when doing things like this.  I don't think it matters much for this project since we're embedding this in go, and go makes heap allocations like its going out of style.  Worth noting tho.

Reply:
I agree, but unfortunately in most places here we need a length prefix for some parts of the encoded data.
So we need to encode the data first to get that length.